api/v1alpha1/composable_webhook.go

```diff
@@ -0,0 +1,271 @@
+				}
+				composablelog.Info("getValueFrom is valid")
+				//set to a random value for dry-run
+				// TODO: set the value to an appropriate type e.g. int, string, etc
+				m[k] = "abc"
+
+			} else { //recursive checking the sub-elements
+				if err := r.findGetValueFrom(mykey, vv); err != nil {
+					allErrs = append(allErrs, err...)
+				}
+			}
+		default:
+			composablelog.Info("findGetValueFrom", "key", mykey, "type-unknown", vv)
+		}
+	}
+	return allErrs
+}
+
+// findGetValueFrom2 functions in the same way as findGetValueFrom above except taking []interface{} as inputs
+// findGetValueFrom2 processes the ararys in the template content for getValueFrom elements
+func (r *Composable) findGetValueFrom2(path *field.Path, m []interface{}) field.ErrorList {
+	var allErrs field.ErrorList
+	for k, v := range m {
+		mykey := path.Child(strconv.Itoa(k))
+		switch vv := v.(type) {
+		case string, int32, int64, float32, float64, bool:
+		case []interface{}:
+			if err := r.findGetValueFrom2(mykey, vv); err != nil {
+				allErrs = append(allErrs, err...)
+			}
+		case map[string]interface{}:
+			if vv["getValueFrom"] != nil {
+				newM, _ := json.Marshal(vv["getValueFrom"])
+				myGetValueFrom := ComposableGetValueFrom{}
+				json.Unmarshal(newM, &myGetValueFrom)
+				composablelog.Info("found getValueFrom", "key", "getValueFrom", "value", vv["getValueFrom"])
+				if err := validateGetValueFrom(mykey.String(), myGetValueFrom); err != nil {
+					composablelog.Info("getValueFrom is INVALID", "detail", err)
+					allErrs = append(allErrs, field.Invalid(mykey.Child("getValueFrom"), r.Name, err.Error()))
+				}
+				composablelog.Info("getValueFrom is valid")
+				// TODO: set a random value of appropriate type for dry-run
+				m[k] = "abc2"
+			} else {
+				if err := r.findGetValueFrom(mykey, vv); err != nil {
+					allErrs = append(allErrs, err...)
+				}
+			}
+		default:
+			composablelog.Info("findGetValueFrom", "key", k, "type-unknown", vv)
+		}
+	}
+	return allErrs
+}
+
+// validateGetValueFrom validates the syntax of input getValueFrom fields
+func validateGetValueFrom(k string, getValueFrom ComposableGetValueFrom) error {
+	var missingItems []string
+	if getValueFrom.Name == "" {
+		missingItems = append(missingItems, "name")
+	}
+	if getValueFrom.Kind == "" {
+		missingItems = append(missingItems, "kind")
+	}
+	if getValueFrom.Path == "" {
+		missingItems = append(missingItems, "path")
+	}
+	if len(missingItems) > 0 {
+		items := array2string(missingItems)
+		return fmt.Errorf("Missing required field(s) - %v", items)
+	}
+	return nil
+}
+
+func array2string(a []string) string {
+	str := ""
+	for _, v := range a {
+		str = v + "," + str
+	}
+	str = strings.TrimRight(str, ",")
+	return str
+}
+
+// dryrun as a means of syntax validation of the template content
+func (r *Composable) dryRun(m map[string]interface{}, op string) *field.Error {
+	composablelog.Info("dryRun", "name", r.Name)
+	newM, _ := json.Marshal(m)
```

Comment:
should we check returned error ?

Reply:
No, we don't. We only forward the errors. 

If dry-run succeeded (no errors from kube-apiserver), we return `nil` and admission control treats it as an approval.  Otherwise, we forward the dry-run errors from kube-apiserver to admission control which will reject the request with the error messages.

Reply:
I meant a possible return error from `json.Marshal(m)`

Reply:
Oh, I see, will add err catch.